state-chain/pallets/cf-elections/src/electoral_systems/tests/exact_value.rs

```diff
@@ -0,0 +1,199 @@
+};
+use codec::{Decode, Encode};
+use frame_support::assert_ok;
+use scale_info::TypeInfo;
+
+type WitnessExactValueWithStorage = ExactValue<MockIdentifier, Vec<u32>, (), MockHook, u32, u64>;
+
+#[derive(Clone, PartialEq, Eq, Debug, Encode, Decode, TypeInfo, PartialOrd, Ord, Default)]
+pub struct MockIdentifier(Vec<u32>);
+
+impl From<Vec<u32>> for MockIdentifier {
+	fn from(value: Vec<u32>) -> Self {
+		Self(value)
+	}
+}
+
+pub const DO_NOT_STORE_ON_CONSENSUS: [u32; 3] = [0, 1, 2];
+
+pub struct MockHook;
+impl ExactValueHook<MockIdentifier, Vec<u32>> for MockHook {
+	type StorageKey = MockIdentifier;
+	type StorageValue = Vec<u32>;
+
+	fn on_consensus(id: MockIdentifier, value: Vec<u32>) -> Option<(MockIdentifier, Vec<u32>)> {
+		(value != DO_NOT_STORE_ON_CONSENSUS.to_vec()).then_some((id, value))
+	}
+}
+
+fn with_default_state() -> TestContext<WitnessExactValueWithStorage> {
+	TestSetup::<WitnessExactValueWithStorage>::default().build()
+}
+
+#[test]
+fn solana_election_result_reference_counting_works() {
+	with_default_state().then(|| {
+		let mock_id: MockIdentifier = vec![1u32, 2u32].into();
+		let id = |umi: u64| ElectionIdentifier::new(umi.into(), ());
+
+		// Start first election and come to consensus.
+		assert_ok!(WitnessExactValueWithStorage::witness_exact_value::<
+			MockAccess<WitnessExactValueWithStorage>,
+		>(mock_id.clone()));
+		assert!(WitnessExactValueWithStorage::take_election_result::<
+			MockAccess<WitnessExactValueWithStorage>,
+		>(mock_id.clone())
+		.is_none());
+
+		MockStorageAccess::set_consensus_status::<WitnessExactValueWithStorage>(
+			id(0),
+			ConsensusStatus::Gained { most_recent: None, new: vec![11, 12] },
+		);
+		assert_ok!(WitnessExactValueWithStorage::on_finalize::<
+			MockAccess<WitnessExactValueWithStorage>,
+		>(vec![id(0)], &()));
+
+		// First successful lookup `takes` the storage value.
+		assert_eq!(
+			WitnessExactValueWithStorage::take_election_result::<
+				MockAccess<WitnessExactValueWithStorage>,
+			>(mock_id.clone()),
+			Some(vec![11, 12])
+		);
+		assert!(WitnessExactValueWithStorage::take_election_result::<
+			MockAccess<WitnessExactValueWithStorage>,
+		>(mock_id.clone())
+		.is_none());
+
+		// Start elections with overlapping and identical identifiers.
+		let first_id: MockIdentifier = vec![1u32, 2u32].into();
+		let overlap_id: MockIdentifier = vec![2u32, 3u32].into();
+
+		assert_ok!(WitnessExactValueWithStorage::witness_exact_value::<
+			MockAccess<WitnessExactValueWithStorage>,
+		>(first_id.clone()));
+		assert_ok!(WitnessExactValueWithStorage::witness_exact_value::<
+			MockAccess<WitnessExactValueWithStorage>,
+		>(overlap_id.clone()));
+		assert_ok!(WitnessExactValueWithStorage::witness_exact_value::<
+			MockAccess<WitnessExactValueWithStorage>,
+		>(first_id.clone()));
+
+		// Vote and come to consensus
+		MockStorageAccess::set_consensus_status::<WitnessExactValueWithStorage>(
+			id(1),
+			ConsensusStatus::Gained { most_recent: None, new: vec![11, 12] },
```

Comment:
when are [11, 12] taken for election id 1?

Reply:
I think we can just set the result without starting the election. Since this is internal test I thought it'd be OK to omit the first step